### PR TITLE
Defer evaluation of return type hints for serializer functions

### PR DESCRIPTION
**Pull request by dmontagu**
Closes https://github.com/pydantic/pydantic/issues/6489

Also fixes the same issue for `PlainSerializer` etc.

skip change file check

Selected Reviewer: @Kludex

### CODE REVIEW

**Comment by dmontagu** [comment]
need to add tests

please update

File: pydantic/functional_serializers.py [hunk 1]
```diff
@@ -39,8 +40,13 @@ def __get_pydantic_core_schema__(
             The Pydantic core schema.
         """
         schema = handler(source_type)
-        return_type = _decorators.get_function_return_type(self.func, self.return_type)
-        return_schema = None if return_type is None else handler.generate_schema(return_type)
+        try:
+            return_type = _decorators.get_function_return_type(
+                self.func, self.return_type, handler.get_types_namespace()
+            )
```
Comment (adriangb) [inline pydantic/functional_serializers.py:46]: Hmm if this is the only place this is being used I'm inclined to make `_get_types_namespace` a private attribute and just `type: ignore` it, only exposing it when someone else asks for it.

File: pydantic/functional_serializers.py [hunk 2]
```diff
@@ -136,7 +147,7 @@ def field_serializer(
 def field_serializer(
     *fields: str,
     mode: Literal['plain', 'wrap'] = 'plain',
-    return_type: Any = None,
+    return_type: Any = PydanticUndefined,
```
Comment (adriangb) [inline pydantic/functional_serializers.py:150]: Good call on this, obvious in retrospect

**Review by adriangb — APPROVED** [review]
Looks good, please add tests. please update

**Comment by dmontagu** [comment]
please review

**Pull request merged**